[Kevin Chua] iP

kevinchua6 opened the pull request:
### Duke
> "Duke is pretty good" - Kevin

Duke frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/yourally2/ip/releases/download/v0.1/ip.jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is __FREE__!

Features:

- [X] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the```main``` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

kevinchua6 committed:
Add Level 1: Greet, Echo, Exit

kevinchua6 committed:
Add Level 2: Add, List

kevinchua6 committed:
Add Level 3: Mark as Done
* Remove newline at the end when printing list

kevinchua6 committed:
refactor: Todo into Task and TodoList into TaskList
* Accommodate the introduction of ToDos

kevinchua6 committed:
Add Level 4: ToDos, Events, Deadlines

kevinchua6 committed:
Add A-TextUiTesting: Automated Text UI Testing
* Add plural and singular based on number of tasks
* Update .gitignore

kevinchua6 committed:
Add Level 6: Delete
* Further encapsulation for task messages in TaskList

kevinchua6 committed:
Add A-Enums: add enums in the form of Commands
* encapsulate errors related to number of arguments inside Command enum
* change to while loop

kevinchua6 committed:
Add Level-5: Handle Errors
* Add DukeException class to handle handle errors

kevinchua6 committed:
Style: minor touchups

kevinchua6 committed:
Add message box

Message boxes make the User Interface more appealing.
However, it has a small visual bug if the number of characters exceed
the message box.
* Add new class Message

kevinchua6 committed:
Fix: IllegalArgumentException is no longer caught for everything

This is so that in the main loop, IllegalArgumentException is no longer
caught for all statements, but only when Command is instantiated.
Thus, we
* Encapsulate initialisation of Command

kevinchua6 committed:
Refactor(Message): use foreach loop, add comment

Foreach loops are cleaner, and comments help to explain the code.

kevinchua6 committed:
Fix(Command): fix arguments not being verified

Verification is now done outside the method because previously it just
verifies the command string itself, not the arguments.

kevinchua6 committed:
Refactor(Duke): abstracted inputLoop

This is to make it more extensible in case of multiple inputLoops.
Also, I did
* add credits to StackOverflow for Overriding enums
* abstract out the plural and singular for Tasks

kevinchua6 committed:
Fix: fix NullPointerException when calling size of ArrayList

Previously size() would give an NPE when size() is run without first
initialising taskArr.

kevinchua6 committed:
Refactor: abstract creation of Deadline and Event into their respective Classes

This is to follow OOP principles more closely: all the behaviour/errors
involved in the creation of Deadline and Event are now in their
respective Classes and no longer in both Command and Duke.

kevinchua6 committed:
Refactor(Command): remove checking input functionality

It makes more sense for the error checking to be all in one place. As
stated before, this is to follow OOP principles.
Also fix bug where size not being updated for taskWord

kevinchua6 committed:
Add saving when adding tasks and loading

* Creates a text file to store the task list
* Storage of deleted tasks are not implemented yet

kevinchua6 committed:
Add Level 7: Save
* Add saving for add, delete, done actions

kevinchua6 committed:
Add formatting for DateTime

kevinchua6 committed:
Add Level 8: Dates and Times
Uses regex to detect if it is a valid string and parses it accordingly.

kevinchua6 committed:
Merge branch 'branch-Level-7'

kevinchua6 committed:
Merge branch 'branch-Level-8'

kevinchua6 committed:
Add Parser class
To abstract out methods related to user commands.
The Ui class is already abstracted as the Message class which prints
messages.

kevinchua6 committed:
Add duke, command, task and util packages

kevinchua6 committed:
Add JUnit Tests
Add tests for Command, Storage, and TaskList

kevinchua6 committed:
Add Javadocs to all methods

kevinchua6 committed:
Follow the Coding Standard
From this point onward, ensure any new code added are compliant with the
given coding standard.

kevinchua6 committed:
Add Level 9: Find
Makes finding tasks that you want easier

kevinchua6 committed:
Merge remote-tracking branch 'origin/branch-A-JavaDoc'

kevinchua6 committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

kevinchua6 committed:
Merge branch 'branch-Level-9'

kevinchua6 committed:
Refactor code for more OOP and align with reqs

* Rename Message class to Ui
* Move tasks folder outside command folder
* Refactor Duke for more OOP

kevinchua6 committed:
Abstract out *Command classes and refactor code

* Fix tests accordingly
* Extract out parsing logic to Parser

kevinchua6 committed:
Update README.md

kevinchua6 committed:
Fix DateTime not parsing both date and time at once

* Fix bug where taskList is written to twice
* Change behaviour such that it only works for a specific format
This is so that there will not be confusing behaviour where non-datetime
words etc are inputted when you do /by or /at

kevinchua6 committed:
Add better error messages when using wrong format dateTime

kevinchua6 committed:
Merge branch 'add-gradle-support'

kevinchua6 committed:
Add checkstyle and correct all checkstyle errors

kevinchua6 committed:
Merge branch 'branch-A-CheckStyle'

bernarduskrishna commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String FILE = "taskList.txt";
+
+    /** Instance variables */
+    private final Storage myStorage;
```
As this is a constant (final), possibly it's a good idea to capitalise the name?

bernarduskrishna commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,82 @@
+    /** Instance variables */
+    private final Storage myStorage;
+    private TaskList taskList;
+    private final Ui ui;
```
Same thing here. Possibly capitalise?

bernarduskrishna commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
Hmm this is not really about style but intuitively when the command is bye, you kinda expect isExit to be true. I don't quite get this part.

kevinchua6 replied:
Good catch HAHA it's a mistake on my part thanks!

bernarduskrishna commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+    private static final String DEADLINE_DELIMITER = "/by";
+
+    // Messages
+    private static final String INVALID_DEADLINE_MESSAGE = "Invalid use of deadline command. "
+            + "Use 'deadline <text> /by <datetime>'";
+    private static final String INVALID_DEADLINE_FORMAT_MESSAGE = "Invalid DateTime format for deadline. "
+            + "DateTime must be in the format of yyyy-MM-dd HHmm (2019-02-01 1800)";
+    private static final String MISSING_DEADLINE_MESSAGE = "Some arguments are missing. "
+            + "Use 'deadline <text> /by <datetime>'";
+    private static final String INVALID_SAVE_MESSAGE = "Deadline save is given in the wrong format";
```
I like how you consider a lot of different cases beyond what is required.

kevinchua6 replied:
Thank you! Thanks for all the thoughtful comments! I appreciate it.

bernarduskrishna commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,99 @@
+        if (inputArr.length != 3) {
+            throw new DukeException(INVALID_SAVE_MESSAGE);
+        }
+        String isDone = inputArr[0].trim();
```
isDone sounds like a boolean variable name.

bernarduskrishna commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * Represents the input of the user.
+ *
```
Hem is this not 1 extra redundant line? I not so sure abt the styling for javadocs haha.

kevinchua6 replied:
You're right! Thanks!

bernarduskrishna reviewed (commented):
In general, I think you did very well! I like the way you abstracted things and how thorough you are in your considerations (edgecases, etc). Good job.

kevinchua6 committed:
Add Level 10: GUI

* Add the bare minimum for the GUI.
Todo: make it more customised and prettier.

nhjryan commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,98 @@
+     * @return An Event object
+     * @throws DukeException An exception thrown according to the message given
+     */
+    public static Event newEvent(String input, boolean isDone) throws DukeException {
```
Perhaps it would be better to change this to createNewEvent to change this to a verb?

nhjryan commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,185 @@
+     * The size of the array storing the Tasks.
+     * @return The size of the array storing the Tasks.
+     */
+    public int size() {
```
This could also be changed to getSize so that your methods are all consistent to the naming conventions (i.e. methods named as verbs)?

kevinchua6 replied:
Thanks for the comments! I'll work on addressing the issues you pointed out.

nhjryan reviewed (commented):
LGTM! The code generally follows all the coding standards and is written very neatly and cleanly. The only issues I could find were probably just the ones related to making all your method names verbs when necessary (feel free to correct me also, I'm not sure if I'm 100% correct also haha), but overall, the code looks really good! :)

kevinchua6 committed:
Make Duke respond to GUI

* Execute returns a string instead of displaying to allow the GUI show
it
* Abstracted out the two DialogBox so have custom behaviour for each

kevinchua6 committed:
Fix exit command

* Make the UI prettier
* Change the pictures

kevinchua6 committed:
Merge branch 'branch-Level-10'

kevinchua6 committed:
Add whitespace checking to input

Before, when you hit enter when there is no content a blank message
would still be sent. Also, whitespace before and after was not sent as
well. This commit fixes this.

kevinchua6 committed:
Add assertions

This is so that we can verify assumptions about invariants made. Such as
todos, deadlines, and events not being null when adding them.

kevinchua6 committed:
Fix bug causing Duke to not properly exit

This was because it was checking the input instead of the response with
the exit message. An easy and quick fix was done.

kevinchua6 committed:
Improve code quality

Code quality is important to ensure that others can read the code.
Therefore it is important to improve code quality. It is also important
to keep the code maintainable.

Thus, things like import order and branching flow was changed to make it
more readable.

kevinchua6 committed:
Merge pull request #1 from yourally2/branch-A-Assertions

Add assertions

kevinchua6 committed:
Merge pull request #2 from yourally2/branch-A-CodeQuality

Improve code quality

kevinchua6 committed:
Add Fixed Duration Tasks

Implementation is done by overriding the 'todo' command such that it
accepts both Todos and TimedTodos.

kevinchua6 committed:
Fix bug where TimedTodos do not load from taskList

* Add javadocs for new methods
* Abstract out Todo delimiter

kevinchua6 committed:
Merge branch 'branch-B-FixedDurationTasks'

kevinchua6 committed:
Add A-CI: Set up CI

kevinchua6 committed:
Clean up Javadocs.

* Also made all error messages go through the UI instead of CLI.
Except for when duke is initialised.

kevinchua6 committed:
Add welcome message on startup

kevinchua6 committed:
Add User Guide

kevinchua6 committed:
Refactor Duke class

* Remove Launcher class and put logic in Duke class
This is so that we can launch directly from the Duke class and not need
a separate class just to launch the GUI.
* Move Main class to gui package
This is to make the application more organised.

kevinchua6 committed:
Fix fonts on Mac not displaying properly

kevinchua6 committed:
Fix Javadocs

kevinchua6 committed:
Set theme jekyll-theme-cayman

kevinchua6 committed:
Update README.md

kevinchua6 committed:
Fix README.md typos

kevinchua6 committed:
Update images to use royalty free ones

kevinchua6 committed:
Update outer README.md
